### ISSUE TITLE
Why are you describing it this way?

### ISSUE THREAD

**Issue by 1m-N00b** [maximbaz/dotfiles#31]
# Background
I asked you a [question](https://github.com/maximbaz/dotfiles/issues/26) before.
I'm struggling to write my own script based on [your install.sh](https://github.com/maximbaz/dotfiles/blob/master/install.sh). As my name suggests, I'm not familiar with many things, and I only know the basic syntax of shell script, so I'm trying to implement it while researching.
However, I don't quite understand the description [here](https://github.com/maximbaz/dotfiles/blob/2f457996af4f39537829662f50e7b4f75655d76d/install.sh#L111).

```
echo -e "\n### Setting up partitions"
umount -R /mnt 2> /dev/null || true
cryptsetup luksClose luks 2> /dev/null || true
```
# My perception
In the logical OR (command1 || command2), if command1 is false but command2 is true, then the overall result is true.
Also, why is it written as a logical OR (command1 || true) which is always true, instead of just command1, when it doesn't exit on error like set -e?

**I'm sorry, English is my second language and I'm using English translation, so I may not be conveying my questions accurately!**

**Comment by maximbaz** [comment]
Hey 😉 👋 

Your understanding  of how `command1 || command2` behave is correct.

The reason I'm using it like this is because with `set -e` the entire script execution will exit as soon as the first command in the script returns an error. So imagine a script executes a different line, say `mkfs.vfat -n "EFI" -F32 "${part_boot}"`, and this command returns non-zero exit code. Something went wrong, so thanks to `set -e` the entire installation aborts, and we can inspect what went wrong and possibly solve some problem before trying again.

However in other cases we _expect_ errors, and we _want_ to ignore them, we don't want installation script to stop. 

`umount -R /mnt` is one of these cases, you might be running installation script a second time (if it failed for some reason for the first time), and `/mnt` is already mounted. In order to start from scratch, we need to unmount it first, so this is why I added `umount -R /mnt`. However, on the very very first run¸ `/mnt` is not mounted yet, so `umount -R /mnt` will fail as there is nothing to unmount, and the entire installation script will abort. In order not to abort the script, I append `|| true`, which tells `set -e` that it should continue with this script regardless if `umount -R /mnt` succeeded or not.

In summary, when you have `set -e`, using syntax `command1 || true` makes the command optional, it is not required to succeed in order to continue with the rest of the script.

`cryptsetup luksClose luks || true` is the same reason, we might have luks opened from the previous installation, and we must close it if we want to retry installation from scratch, but we don't know if we need to close luks or not, so for simplicity I just marked that command as optional with `|| true`.

Let me know if this makes sense!

**Comment by 1m-N00b** [comment]
I see. I understand now. 
Is this correct in my perception? If set -e is not set, wouldn't the script continue even if umount -R mnt 2> devnull fails? On the other hand, if set -e is set but command1 || true, you can proceed regardless of whether command1 is true or false.
Why do I have to do a logical OR without set -e?

**It may be my lack of knowledge or awareness.**

**Comment by maximbaz** [comment]
> If set -e is not set, wouldn't the script continue even if umount -R mnt 2> devnull fails?

Correct 👍 

> On the other hand, if set -e is set but command1 || true, you can proceed regardless of whether command1 is true or false.

Correct 👍 

> Why do I have to do a logical OR without set -e?

Aaah this is what confused you! Sorry, I forgot about this bit of magic as well 😄 

It's because instead of `set -e` I have the line below, which is basically doing exactly the same as `set -e`, but additionally it prints what command failed:

```
trap 's=$?; echo "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $s' ERR
```

You can try running these scripts (`htop -a` fails because it's invalid flag):

```sh
trap 's=$?; echo "$0: Error on line "$LINENO": $BASH_COMMAND"; exit $s' ERR

htop -a
echo success
```


```sh
set -e

htop -a
echo success
```

**Comment by 1m-N00b** [comment]
I see. I understand perfectly. Thank you for making it so easy to understand.
Please have two more questions for you.

## Q1
Also, this is a different question, but why do you use wayland?
I know it has better security while you can't share your screen with discord etc. and wayland have some issues.

## Q2
Do you have [pkgs, aurbuild, and archbuild on btrfs](https://github.com/maximbaz/dotfiles/blob/2f457996af4f39537829662f50e7b4f75655d76d/install.sh#L138-L140) because you are using aurutils instead of yay? Also, do you use aurutils because you are TU and manage package? I thought yay(pacman wrapper) was fine for personal use.

**Comment by maximbaz** [comment]
Happy to answer 😉 

### Q1

I use Wayland because it's being designed and developed with modern needs in mind. I'll try to come up with a few examples of the top of my mind:

- I love HiDPI screens, and so when you have two screens whose DPI values don't match (e.g. laptop screen and external monitor, or two external monitors of different brands, doesn't matter), you are out of luck on Xorg, there is simply no way to configure two outputs to have different DPI values. There are many hacks and believe me I tried them all, nothing works without trouble. Wayland on the other hand has been designed with an understanding that people can have multiple outputs with different resolutions and DPI values, and this all _just works_ out of the box. By the way, earlier this year Gnome released some significant improvements to this for Xorg, but it's still not perfect and buggy.
- It might be subjective but I think I even see a difference in quality when I use a single 4k screen, on Wayland everything is rendered more sharp than on Xorg, I don't know why, but on Xorg it feels a bit more blurry.
- Not only outputs (monitors) are easier to work with in Wayland, but inputs as well (mouse, keyboard, touchpad). For example, in kitty terminal you can achieve [high precision accelerated touchpad scrolling](https://sw.kovidgoyal.net/kitty/conf.html#opt-kitty.touch_scroll_multiplier) only on Wayland - I don't remember if this is due to buggy drivers or something else, but anyway it's a fact.

It is true that some apps work better on Xorg, but it's not Wayland's fault, the functionality is available, but apps need to make use of it, so it depends on app developers. For example screen sharing is possible and works well via WebRTC in Chromium-based browsers, and as far as I know it works in Electron 12 (which is in beta until March), so after March screen sharing in all electron-based apps (which I think includes discord) will start working too. 

I managed to get an environment that satisfied all my needs, I replaced apps that work only in Xorg with some Wayland-based alternatives, and am happy ever since 😉 

### Q2

- `pkgs` is a volume for `pacman` cache, you need it even if you don't use AUR at all. It's needed so that when you take system snapshot, this cache will not be part of snapshot (because it is huge and can always be restored from archive.archlinux.org).
- `archbuild` is a volume for `devtools`, needed for TU activities, for building packages in clean environment using e.g. `extra-x86_64-build`. This you can skip unless you also use `devtools` for some other purposes.
- `aurbuild` is a volume for `aurutils`.

> Also, do you use `aurutils` because you are TU and manage package? I thought `yay` (pacman wrapper) was fine for personal use.

Not at all, I used `aurutils` even before becoming TU 🙂 `yay` is fine, and I used it in the past, before I discovered `aurutils` actually. I like `aurutils` because contrary to all other helpers it doesn't replace `pacman`, but it complements it. It taught me a lot about how packaging works in general in Arch Linux, how `pacman` works, what are those repositories, etc. You begin to manage your own system like TUs manage official package, which I think is very cool. Also I particularly enjoy that with `aurutils` I can build every AUR package in a clean isolated environment (just like with official packages), free of unexpected undesired side-effects, so if you and I build the same package we will get the same result, regardless of what other packages we have installed on our systems.

Hope this answers your questions, do not hesitate to ask more 😉

**Comment by 1m-N00b** [comment]
Thank you for your detailed and easy-to-understand answer with concrete examples.
I gained useful knowledge that I did not have in my knowledge.thank you very much

Good luck with your daily work👍

**Comment by maximbaz** [comment]
@1m-N00b on the question of why to use Wayland, you might find [this blog post](https://ajaxnwnk.blogspot.com/2020/10/on-abandoning-x-server.html) interesting, by the Xorg maintainer.

**Comment by 1m-N00b** [comment]
Thank you very much for coming to tell me all the way
